### PR TITLE
APP-506: Offer loading state

### PR DESCRIPTION
**Pull request by hugokallstrom**
### Checklist

- [ ] Functionality is covered by an integration test
- [x] Functionality is accessible in engineering mode

### CODE REVIEW

File: app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt [hunk 1]
```diff
@@ -127,20 +132,27 @@ class OfferActivity : BaseActivity(R.layout.activity_offer) {
...
-                when (viewState) {
-                    OfferViewModel.ViewState.HasContracts -> startLoggedInActivity()
-                    is OfferViewModel.ViewState.OfferItems -> {
-                        topOfferAdapter.submitList(viewState.topOfferItems)
-                        perilsAdapter.submitList(viewState.perils)
-                        insurableLimitsAdapter.submitList(viewState.insurableLimitsItems)
-                        documentAdapter.submitList(viewState.documents)
-                        bottomOfferAdapter.submitList(viewState.bottomOfferItems)
+            lifecycleScope.launchWhenResumed {
+                model.viewState.collect { viewState ->
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt:136]: In order to 'mimic' the desirable `Lifecycle`-behaviour of `LiveData` (collect when > `onResume`, pause in `onPause`, restart collection in `onResume`, etc), we should use `flowWithLifecycle` to collect this flow.

```suggestion
                model.viewState.flowWithLifecycle(this, Lifecycle.State.STARTED).collect { viewState ->
```

Here's an article on the subject: https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda 🙂

File: app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt [hunk 2]
```diff
@@ -127,20 +132,27 @@ class OfferActivity : BaseActivity(R.layout.activity_offer) {
...
+                    when (viewState) {
+                        OfferViewModel.ViewState.HasContracts -> startLoggedInActivity()
+                        is OfferViewModel.ViewState.OfferItems -> {
+                            topOfferAdapter.submitList(viewState.topOfferItems)
+                            // Workaround - recyclerview will scroll to bottom if updating all items simultaneously.
+                            lifecycleScope.launch {
+                                delay(500)
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt:143]: Interesting 🤔 is 500 ms required, or just a number that works?

I guess as long as we do it after the `RecyclerView` has finished laying out the new items, we should be good, right? Perhaps we could use `doOnNextLayout` and post the additional items from that 🙂

File: app/src/main/res/layout/activity_offer.xml [hunk 3]
```diff
@@ -13,7 +13,7 @@
         android:layout_width="match_parent"
         android:layout_height="match_parent"
         android:clipToPadding="false"
-        android:paddingBottom="120dp"
+        android:layout_above="@+id/signButton"
```
Comment (hedvigoscar) [inline app/src/main/res/layout/activity_offer.xml:16]: Whoa, does this work? 🤯 I thought `layout_above` was a `RelativeLayout`-only attribute? 🤔

**Review by hedvigoscar — COMMENTED** [review]
Nice!

File: app/src/main/java/com/hedvig/app/feature/offer/OfferViewModel.kt [hunk 4]
```diff
@@ -51,6 +53,7 @@ abstract class OfferViewModel : ViewModel() {
             object EmptyResponse : Error()
         }
 
+        class Loading(val loadingItem: List<OfferModel.Loading>) : ViewState()
```
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/offer/OfferViewModel.kt:56]: Agree, fixing

File: app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt (hunk 1 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt:136]: Nice

File: app/src/main/res/layout/activity_offer.xml (hunk 3 above)
Comment (hugokallstrom) [reply app/src/main/res/layout/activity_offer.xml:16]: Thought so too, you learn something new everyday :D

File: app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt (hunk 2 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt:143]: Good idea, that works great.

**Review by hedvigoscar — APPROVED** [review]

**Pull request merged**